### PR TITLE
[Docs] Add developer documentation for using/modifying the chrome service

### PR DESCRIPTION
**Pull request by kaddy645**
Signed-off-by: kaddy645 <xdeskart@amazon.com>

### Description
As a plugin or core dashboards developer, I'd like to have a README for the [chrome service](https://github.com/opensearch-project/OpenSearch-Dashboards/tree/main/src/core/public/chrome) to be able to quickly and consistently integrate plugins with the global application chrome.

### SC
Screen Shot 2022-07-11 at 10 57 38 AM

### Issues Resolved
[Issue-1857](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1857)
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
    - [ ] `yarn test:jest`
    - [ ] `yarn test:jest_integration`
    - [ ] `yarn test:ftr`
- [ ] New functionality has been documented.
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/core/public/chrome/README.md [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    chrome.recentlyAccessed.get().forEach(console.log);;
+   ```
+   
+### DocTitleService : 
+- Interface : ChromeDocTitle
+- Signature : ```docTitle: ChromeDocTitle```
+  ###### - How to change the title of the document
+
+
+   ```ts
```
Comment (ashwin-pc) [inline src/core/public/chrome/README.md:125]: These snippets are very useful, thank you!

File: src/core/public/chrome/README.md [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+   chrome.docTitle.change('My application title')
+   chrome.docTitle.change(['My application', 'My section'])
+   ```
+### UI :
+###### consists of tsx/scss files && renders UI components from css Library e.g ```<Progress props={props}>```
+
+- Adding/overriding existing css : e.g.
+  ```jsx
```
Comment (ashwin-pc) [inline src/core/public/chrome/README.md:133]: Can we limit the code snippet markdown to just code examples? Its a lil unreadable when the snippet markdown is used for general text :(

File: src/core/public/chrome/README.md [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+       
+       - truncate: boolean Forces all breadcrumbs to single line and truncates each breadcrumb to a particular width, except for the last item
+       
+	   - max : Collapses the inner items past the maximum set here into a single ellipses item.
+	   
+	   - breadcrumbs : The array of individual Breadcrumb items
+       ```
+### Custom React Hooks :
+<b> useObservable</b> : React state hook that tracks the latest value of an Observable.
+[More About "React-use" custom hooks library](https://github.com/streamich/react-use)
```
Comment (ashwin-pc) [inline src/core/public/chrome/README.md:151]: Do we expose only the `useObservable` hook from `streamich/react-use`? or have we manually copied over the code from that library?

**Review by ashwin-pc — COMMENTED** [review]
Always lovely to see new docs! 

Love the work put in to make it look presentable, but i think that the lack of standardization in markdown is kinda ruining it a little. Given that we dont know what viewer the user will use to read this file, lets make sure it semantically correct? I'm open to suggestions but imo i'd:
- avoid using html in markdown
- use \``` only for multiline code snippets
- use code snippet blocks strictly for code (The code snippet should be copy paste-able)
- make sure that the readme is easily readable without a markdown viewer too

As for the doc itself, the content is good. I'd only like to see more of an abstract of what the service is and does as a whole so that i can decide if its what i want.

File: src/core/public/chrome/README.md (hunk 3 above)
Comment (kaddy645) [reply src/core/public/chrome/README.md:151]: At the moment ,we are only exposing useObservable.  I can see react-use in package json!

File: src/core/public/chrome/README.md (hunk 2 above)
Comment (kaddy645) [reply src/core/public/chrome/README.md:133]: Makes sense!

File: src/core/public/chrome/README.md [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+## About :
+- <b> Signature </b> - ```export interface ChromeStart```
+- <b> CoreStart</b> - Core services exposed to the Plugin start lifecycle.
+- <b> How to access ? </b>  add in interface ```chrome: ChromeStart && chrome.servicesName => e.g chrome.docTitle.method```
+- <b> Where it is getting Registered/Executed </b> [Staring Point](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/public/rendering/rendering_service.tsx)
+- <b> How header component is getting rendered </b> ``` const chromeUi = chrome.getHeaderComponent(); ```
+- <b> Chrome Methods </b> [See chrome interface/methods](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/53e98a66a81cf986894962b5f17e4190baebf57f/src/core/public/chrome/chrome_service.tsx)
+- <b> Consists of these sub-services/components </b> : 
+```js
```
[the remark this answers is not in the document]
Comment (kaddy645) [reply src/core/public/chrome/README.md:35]: Yes!

File: src/core/public/chrome/README.md [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+- [Nav Links Service](#navlinksservice-)
+- [Recently Accessed Service](#recentlyaccessedservice-)
+- [Doc Title Service](#doctitleservice-)
+- [UI](#ui-)
+- [Custom React Hooks](#custom-react-hooks-)
+- [Common RX Js Terms and Operator](#common-rx-js-terms-and-operator-)
+
+
+## About :
+- <b> Signature </b> - ```export interface ChromeStart```
```
[the remark this answers is not in the document]
Comment (kaddy645) [reply src/core/public/chrome/README.md:28]: makes sense!

**Comment by CPTNB** [comment]
Thanks for writing this up!  I see a few different types of content in this file and I'd like you to tease them into individual files and places.

At the highest level you have general guidance about what the plugin is, why it exists, how it works, and in vague terms how it's used.  This is good stuff as an at-a-glance reference to future developers, and I think it's what Josh had in mind when he wrote the original issue.

Next you have more lower level API documentation, and I'm not so excited to see this kind of stuff in this type of document.  Personally, I prefer my API docs to be very well structured so that I can quickly scan them to get the information I need.  I don't want to scan through a narrative to fish out API info.  I'm not sure hand-written API docs is the right way to go anyway, since that sort of thing is easily (at least, more easily) auto-generated and automatically kept up to date.

Last, you have some information about libraries and concepts like Observables.  I do think this kind of thing has a place, but I'd like to see it live elsewhere from plugin-specific information.

My guidance would be to coalesce the high level information about the plugin into a few narrative paragraphs at the top of the document, then group the API docs into task driven examples that show code blocks that solve what you think might be common tasks.  Then, move the information about libraries and Observables to the more free-form blog docs: https://github.com/CPTNB/opensearch-dashboards-dev-docs

**Review by ashwin-pc — DISMISSED** [review]
Amazing! thanks

File: src/core/public/chrome/README.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+- [Nav Controls Service](#navcontrolsservice-)
+- [Nav Links Service](#navlinksservice-)
+- [Recently Accessed Service](#recentlyaccessedservice-)
+- [Doc Title Service](#doctitleservice-)
+- [UI](#ui-)
+
+## About :
+- **Signature** - `export interface ChromeStart`
+
+The chrome service is a high level UI service that is part of CoreStart (Core services exposed to the Plugin start lifecycle) and offers other plugins a way to add navigation controls to the UI, edit the document title, manipuate navlinks on global header as well as edit the Recent accessed tab. it consists of these sub-services/components.
```
Comment (ananzh) [inline src/core/public/chrome/README.md:14]: This part looks like an overall intro to me.
```
The chrome service is a high level UI service that is part of CoreStart (Core services exposed to the Plugin start lifecycle) and offers other plugins a way to add navigation controls to the UI, edit the document title, manipuate navlinks on global header as well as edit the Recent accessed tab. it consists of these sub-services/components.
```

Right now, display this intro in between `Signature` and `NavControlsService` looks really strange to me. 
Screen Shot 2022-07-15 at 15 04 42

what do you think?

Comment (ananzh) [reply src/core/public/chrome/README.md:14]: Nit: 
`Recent` --> `recent`
`it consists` --> `It consists`

File: src/core/public/chrome/README.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+- NavControlsService : for registering new controls to be displayed in the navigation bar.
+- NavLinksService : for manipulating nav links.
+- RecentlyAccessedService : for recently accessed history.
+- DocTitleService: for accessing and updating the document title
+- UI : All the UI components,icons for e.g  header, loaders.
```
Comment (ananzh) [inline src/core/public/chrome/README.md:19]: I think `e.g.` already means `for example` right?
should we delete `for` ?

File: src/core/public/chrome/README.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+
+- How to access ? add in interface `chrome: ChromeStart && chrome.servicesName => e.g chrome.docTitle.method`
```
Comment (ananzh) [inline src/core/public/chrome/README.md:22]: `e.g` --> `e.g.`

File: src/core/public/chrome/README.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    return () => ReactDOM.unmountComponentAtNode(targetDomElement);
+  }
+})
+```
+### NavLinksService : 
+- Interface :  ChromeNavLinks
+- **Signature** - `navLinks: ChromeNavLinks`
+- e.g Method : 
+
+Get an observable for a sorted list of navlinks :-
```
Comment (ananzh) [inline src/core/public/chrome/README.md:49]: Can we make formats more consistent? For example, can we format these two parts same?

Screen Shot 2022-07-15 at 15 14 02

File: src/core/public/chrome/README.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+The chrome service is a high level UI service that is part of CoreStart (Core services exposed to the Plugin start lifecycle) and offers other plugins a way to add navigation controls to the UI, edit the document title, manipuate navlinks on global header as well as edit the Recent accessed tab. it consists of these sub-services/components.
+- NavControlsService : for registering new controls to be displayed in the navigation bar.
+- NavLinksService : for manipulating nav links.
+- RecentlyAccessedService : for recently accessed history.
+- DocTitleService: for accessing and updating the document title
+- UI : All the UI components,icons for e.g  header, loaders.
+
+
+- How to access ? add in interface `chrome: ChromeStart && chrome.servicesName => e.g chrome.docTitle.method`
+- Where it is getting Registered/Executed [Staring Point](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/2.1/src/core/public/rendering/rendering_service.tsx)
```
Comment (ananzh) [inline src/core/public/chrome/README.md:23]: Can we make the format more consistent?
add question mark after each question (L23 and L24), just like what you did for the first one `How to access?`?

File: src/core/public/chrome/README.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+### RecentlyAccessedService : 
+
+- Interface : ChromeRecentlyAccessed
+- Signature : ```recentlyAccessed: ChromeRecentlyAccessed```
+- The Recently viewed items are stored in the browser's local storage.
+- You can go back to the recent search/visualization/dashboard , each item has (link,label,id)
+- Methods :
+
+Adds a new item to the recently accessed history :-
```
Comment (ananzh) [inline src/core/public/chrome/README.md:83]: Same format issue. Needs some consistency.

**Review by ananzh — CHANGES_REQUESTED** [review]

**Review by ananzh — APPROVED** [review]
NIT> some `e.g` left in the doc.

**Pull request merged**